src/decoder/cuda-decoder-utils.h

```diff
@@ -0,0 +1,161 @@
+// decoder/cuda-decoder-utils.h
+
+// Copyright      2018  Zhehuai Chen
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef KALDI_CUDA_DECODER_UTILS_H_
+#define KALDI_CUDA_DECODER_UTILS_H_
+
+
+#include <cuda_runtime_api.h>
+#include "cuda_runtime.h"
+#include "util/stl-utils.h"
+#include "omp.h"
+#include <algorithm>
+#include <float.h>
+#include <math.h>
+#include <cooperative_groups.h>
+#include "math_constants.h"
+#include "omp.h"
+
+// cuda macro
+
+#ifdef __CUDACC__
+#define HOST __host__
+#define DEVICE __device__
+
+#else
+#define HOST
+#define DEVICE
+#endif
+
+#define USE_NVTX
+#ifdef USE_NVTX
+#include "nvToolsExt.h"
+const uint32 colors[] = {0x0000ff00, 0x000000ff, 0x00ffff00, 0x00ff00ff, 
+                         0x0000ffff, 0x00ff0000, 0x00ffffff};
+const int32 num_colors = sizeof(colors) / sizeof(uint32);
+
+#define PUSH_RANGE(name,cid) { \
+    int32 color_id = cid; \
+    color_id = color_id%num_colors;\
+    nvtxEventAttributes_t eventAttrib = {0}; \
+    eventAttrib.version = NVTX_VERSION; \
+    eventAttrib.size = NVTX_EVENT_ATTRIB_STRUCT_SIZE; \
+    eventAttrib.colorType = NVTX_COLOR_ARGB; \
+    eventAttrib.color = colors[color_id]; \
+    eventAttrib.messageType = NVTX_MESSAGE_TYPE_ASCII; \
+    eventAttrib.message.ascii = name; \
+    nvtxRangePushEx(&eventAttrib); \
+}
+#define POP_RANGE nvtxRangePop();
+#else
+#define PUSH_RANGE(name,cid)
+#define POP_RANGE
+#endif
+
+// Macro for checking cuda errors following a cuda launch or api call
+#define cudaCheckError() {                                          \
```

Comment:
Is there a problem with reusing the macro defined somewhere under the cumatrix/ directory?

Reply:
in other parts of kaldi, we use some kind like this:
CU_SAFE_CALL(cudaGetLastError());